### PR TITLE
Update Memo structure to latest specification

### LINKED ISSUE #199: Refactor Memo to follow the specification
`Memo` should be

```
memo = [ + (tstr / (bstr .cbor any)) ]
```

See specification in [1]

[1]: https://github.com/many-protocol/specification/blob/f1a2bad7ee9413d4ec6c068660e2366e05c342e6/spec/cddl/types.cddl#L41-L47

### PR DESCRIPTION
**Pull request by hansl**
Fixes #199

### CODE REVIEW

**Comment by hansl** [comment]
Explanation of the latest 2 commits; for some reason `cargo install` started failing and I had to upgrade the rust toolchain to 1.63 (I think `cargo install` will use latest stable?). Then new clippy errors started popping so I had to fix those.

File: src/many-modules/src/_9_account/features/multisig.rs [hunk 1]
```diff
@@ -9,98 +9,168 @@ use many_macros::many_module;
...
+            return Err(ManyError::unknown(format!(
+                "Data size ({}) over limit ({})",
+                b.len(),
+                M
+            )));
         }
-        Ok(Data(b.into()))
+        Ok(Self {
+            inner: vec![MemoInner::ByteString(b)],
+        })
```
Comment (fmorency) [inline src/many-modules/src/_9_account/features/multisig.rs:153]: Isn't that already checked by the `MemoInner`, i.e., can't you simply do

```
Ok(Self::from(MemoInner::try_from(b)?))
```

File: src/many-modules/src/_9_account/features/multisig.rs [hunk 2]
```diff
@@ -9,98 +9,168 @@ use many_macros::many_module;
...
+/// type that requires meta information.
+#[derive(Default, Clone, Debug, PartialOrd, Eq, PartialEq)]
+pub struct Memo<const MAX_LENGTH: usize = MULTISIG_MEMO_DATA_MAX_SIZE> {
+    inner: Vec<MemoInner<MAX_LENGTH>>,
+}
+
+impl<const M: usize> Memo<M> {
+    pub fn new() -> Self {
+        Self::default()
+    }
```
Comment (fmorency) [inline src/many-modules/src/_9_account/features/multisig.rs:102]: Would this allow the creation of a `Memo` with an empty `inner`?

We need to enforce that `The memo itself CANNOT be empty and MUST contain at least one entry.`

**Review by fmorency — COMMENTED** [review]
Thanks!

Some questions before approval.

File: src/many-modules/src/_9_account/features/multisig.rs (hunk 1 above)
Comment (hansl) [reply src/many-modules/src/_9_account/features/multisig.rs:153]: You're right. I did the `String` version but not this one. My bad.

File: src/many-modules/src/_9_account/features/multisig.rs (hunk 2 above)
Comment (hansl) [reply src/many-modules/src/_9_account/features/multisig.rs:102]: Removed `Default` and `new()`.

**Review by fmorency — APPROVED** [review]

**Review by Denommus — APPROVED** [review]

**Comment by hansl** [comment]
@fmorency if you have time, I added a backward compatible decoding so we can make a migration in framework.

**Comment by hansl** [comment]
I'll block merging this PR for now (moving to Draft) so that we can merge it when the migration in framework is ready.

**Comment by fmorency** [comment]
> @fmorency if you have time, I added a backward compatible decoding so we can make a migration in framework.

I will look at it after the migration in framework is ready.

File: src/many-modules/src/_9_account/features/multisig.rs [hunk 3]
```diff
@@ -9,112 +9,189 @@ use many_macros::many_module;
...
+impl<const M: usize> From<MemoInner<M>> for Memo<M> {
+    fn from(inner: MemoInner<M>) -> Self {
+        Self { inner: vec![inner] }
     }
 }
 
-/// Data inside a transaction
-#[derive(Clone, Debug, Eq, PartialEq)]
-pub struct Data(ByteVec);
+impl<const M: usize> TryFrom<Either<String, ByteVec>> for Memo<M> {
```
Comment (Denommus) [inline src/many-modules/src/_9_account/features/multisig.rs:131]: Why is this a `TryFrom` and not a `From` if there's no failure path?

File: src/many-modules/src/_9_account/features/multisig.rs [hunk 4]
```diff
@@ -9,112 +9,189 @@ use many_macros::many_module;
...
-                s.as_str().as_bytes().len()
-            ));
-        }
-        Ok(Memo(s))
+/// A memo contains a human readable portion and/or a machine readable portion.
+/// It is meant to be a note regarding a message, transaction, info or any
+/// type that requires meta information.
+#[derive(Clone, Debug, PartialOrd, Eq, PartialEq)]
+pub struct Memo<const MAX_LENGTH: usize = MULTISIG_MEMO_DATA_MAX_SIZE> {
+    inner: Vec<MemoInner<MAX_LENGTH>>,
```
Comment (Denommus) [inline src/many-modules/src/_9_account/features/multisig.rs:96]: This is wrong. Memo is a `[ + (tstr / (bstr .cbor any)) ]`, which means there must be at least one element. You need another field `first: MemoInner<MAX_LENGTH>` and adapt the code accordingly. This way you're preventing an empty `Memo` in the type level.

File: src/many-modules/src/_9_account/features/multisig.rs [hunk 5]
```diff
@@ -9,112 +9,189 @@ use many_macros::many_module;
...
+        _: &mut C,
+    ) -> Result<(), encode::Error<W::Error>> {
+        e.encode(&self.inner).map(|_| ())
     }
+}
 
-    pub fn as_bytes(&self) -> &[u8] {
-        self.0.as_slice()
+impl<'b, C, const M: usize> Decode<'b, C> for Memo<M> {
+    fn decode(d: &mut Decoder<'b>, ctx: &mut C) -> Result<Self, decode::Error> {
```
Comment (Denommus) [inline src/many-modules/src/_9_account/features/multisig.rs:176]: It's possible to decode an empty `Memo` from here, which, as previously stated, is a wrong behavior according to the specification.

**Comment by hansl** [comment]
Blocked on https://github.com/liftedinit/many-framework/pull/243 (because we need migrations).

File: src/many-modules/src/_9_account/features/multisig.rs (hunk 5 above)
Comment (hansl) [reply src/many-modules/src/_9_account/features/multisig.rs]: Done and added a test.

File: src/many-modules/src/_9_account/features/multisig.rs (hunk 4 above)
Comment (hansl) [reply src/many-modules/src/_9_account/features/multisig.rs]: I'm not going to add a new type for `NeverEmptyVec` (or add a crate). It's a simple invariant to test for.

**Comment by Denommus** [comment]
CI failing

File: src/many-modules/Cargo.toml [hunk 6]
```diff
@@ -21,10 +21,13 @@ strum = "0.24.1"
 strum_macros = "0.24.2"
 
 [dev-dependencies]
-cbor-diag = "0.1.9"
+cbor-diag = "0.1.10"
```
Comment (fmorency) [inline src/many-modules/Cargo.toml:24]:
```suggestion
cbor-diag = "0.1.11"
```

File: src/many-modules/src/_4_events.rs [hunk 7]
```diff
@@ -962,7 +960,8 @@ mod test {
                                 threshold: None,
                                 timeout_in_secs: None,
                                 execute_automatically: None,
-                                data: None,
+                                data_: None,
+                                memo_: None,
```
Comment (fmorency) [inline src/many-modules/src/_4_events.rs:964]: nit: Impl sane `defaults()`

File: src/many-types/Cargo.toml [hunk 8]
```diff
@@ -19,5 +19,9 @@ proptest = { version = "1.0.0", optional = true }
 serde = "1.0.139"
 
 [dev-dependencies]
+cbor-diag = "0.1.10"
```
Comment (fmorency) [inline src/many-types/Cargo.toml:22]:
```suggestion
cbor-diag = "0.1.11"
```

File: src/many-types/src/memo/legacy.rs [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    // USED ONLY IN TESTS! NEVER REMOVE THIS GUARD! MEMOS SHOULDN'T BE PRODUCED FROM LARGE STRINGS
+    #[cfg(test)]
+    pub fn new(s: String) -> Self {
+        Memo(s)
+    }
+}
+
+/// Data inside a transaction
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Data(ByteVec);
```
Comment (fmorency) [inline src/many-types/src/memo/legacy.rs:66]: I don't think this code should not be located in this file

File: src/many-types/src/memo.rs [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+    proptest! {
+        #[test]
+        fn memo_str_decode_prop(len in 900..1100usize) {
+            let data = String::from_utf8(vec![b'A'; len]).unwrap();
+            let cbor = format!(r#" [ "{data}" ] "#);
+            let bytes = cbor_diag::parse_diag(cbor).unwrap().to_bytes();
+
+            let result = minicbor::decode::<Memo<1000>>(&bytes);
+            if len <= 1000 {
+                assert!(result.is_ok());
```
Comment (fmorency) [inline src/many-types/src/memo.rs:209]: Check the integrity of the data

File: src/many-types/src/memo.rs [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+
+        #[test]
+        fn memo_bytes_decode_prop(len in 900..1100usize) {
+            let data = hex::encode(vec![1u8; len]);
+            let cbor = format!(r#" [ h'{data}' ] "#);
+            let bytes = cbor_diag::parse_diag(cbor).unwrap().to_bytes();
+
+            let result = minicbor::decode::<Memo<1000>>(&bytes);
+            if len <= 1000 {
+                assert!(result.is_ok());
```
Comment (fmorency) [inline src/many-types/src/memo.rs:223]: Check the integrity of the data

File: src/many-types/src/memo.rs [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+        }
+    }
+
+    #[test]
+    fn memo_decode_ok() {
+        let data = String::from_utf8(vec![b'A'; MEMO_DATA_DEFAULT_MAX_SIZE]).unwrap();
+        let cbor = format!(r#" [ "{data}" ] "#);
+        let bytes = cbor_diag::parse_diag(cbor).unwrap().to_bytes();
+
+        assert!(minicbor::decode::<Memo>(&bytes).is_ok());
```
Comment (fmorency) [inline src/many-types/src/memo.rs:236]: Check the integrity of the data

File: src/many-types/src/memo.rs [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+    }
+
+    #[test]
+    fn memo_decode_too_large() {
+        let data = String::from_utf8(vec![b'A'; MEMO_DATA_DEFAULT_MAX_SIZE + 1]).unwrap();
+        let cbor = format!(r#" [ "{data}" ] "#);
+        let bytes = cbor_diag::parse_diag(cbor).unwrap().to_bytes();
+
+        assert!(minicbor::decode::<Memo>(&bytes).is_err());
```
Comment (fmorency) [inline src/many-types/src/memo.rs:245]: Check the error type

File: src/many-types/src/memo.rs [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+    }
+
+    #[test]
+    fn memo_decode_empty() {
+        let cbor = " [] ";
+        let bytes = cbor_diag::parse_diag(cbor).unwrap().to_bytes();
+
+        assert!(minicbor::decode::<Memo>(&bytes).is_err());
```
Comment (fmorency) [inline src/many-types/src/memo.rs:253]: Check the error type

File: src/many-types/src/memo.rs [hunk 15]
```diff
@@ -0,0 +1,365 @@
...
+    }
+
+    #[test]
+    fn data_decode_ok() {
+        let data = hex::encode(vec![1u8; MEMO_DATA_DEFAULT_MAX_SIZE]);
+        let cbor = format!(r#" [ h'{data}' ] "#);
+        let bytes = cbor_diag::parse_diag(cbor).unwrap().to_bytes();
+
+        assert!(minicbor::decode::<Memo>(&bytes).is_ok());
```
Comment (fmorency) [inline src/many-types/src/memo.rs:262]: Check the integrity of the data

File: src/many-types/src/memo.rs [hunk 16]
```diff
@@ -0,0 +1,365 @@
...
+    #[test]
+    fn memo_decode_empty() {
+        let cbor = " [] ";
+        let bytes = cbor_diag::parse_diag(cbor).unwrap().to_bytes();
+
+        assert!(minicbor::decode::<Memo>(&bytes).is_err());
+    }
+
+    #[test]
+    fn data_decode_ok() {
```
Comment (fmorency) [inline src/many-types/src/memo.rs:257]:
```suggestion
    fn bytes_decode_ok() {
```
nit

File: src/many-types/src/memo.rs [hunk 17]
```diff
@@ -0,0 +1,365 @@
...
+        let data = hex::encode(vec![1u8; MEMO_DATA_DEFAULT_MAX_SIZE]);
+        let cbor = format!(r#" [ h'{data}' ] "#);
+        let bytes = cbor_diag::parse_diag(cbor).unwrap().to_bytes();
+
+        assert!(minicbor::decode::<Memo>(&bytes).is_ok());
+    }
+
+    #[test]
+    fn data_decode_large() {
```
Comment (fmorency) [inline src/many-types/src/memo.rs:266]:
```suggestion
    fn bytes_decode_large() {
```
nit

File: src/many-types/src/memo.rs [hunk 18]
```diff
@@ -0,0 +1,365 @@
...
+        let data = hex::encode(vec![1u8; MEMO_DATA_DEFAULT_MAX_SIZE + 1]);
+        let cbor = format!(r#" [ h'{data}' ] "#);
+        let bytes = cbor_diag::parse_diag(cbor).unwrap().to_bytes();
+
+        assert!(minicbor::decode::<Memo>(&bytes).is_err());
```
Comment (fmorency) [inline src/many-types/src/memo.rs:271]: Check the error type

File: src/many-types/src/memo.rs [hunk 19]
```diff
@@ -0,0 +1,365 @@
...
+    }
+
+    #[test]
+    fn mixed_decode_ok() {
+        let data = hex::encode(vec![1u8; MEMO_DATA_DEFAULT_MAX_SIZE]);
+        let cbor = format!(r#" [ "", h'{data}', "" ] "#);
+        let bytes = cbor_diag::parse_diag(cbor).unwrap().to_bytes();
+
+        assert!(minicbor::decode::<Memo>(&bytes).is_ok());
```
Comment (fmorency) [inline src/many-types/src/memo.rs:280]: Check the integrity of the data

File: src/many-types/src/memo.rs [hunk 20]
```diff
@@ -0,0 +1,365 @@
...
+    }
+
+    #[test]
+    fn mixed_decode_data_too_lare() {
+        let data = hex::encode(vec![1u8; MEMO_DATA_DEFAULT_MAX_SIZE + 1]);
+        let cbor = format!(r#" [ "", h'{data}', "" ] "#);
+        let bytes = cbor_diag::parse_diag(cbor).unwrap().to_bytes();
+
+        assert!(minicbor::decode::<Memo>(&bytes).is_err());
```
Comment (fmorency) [inline src/many-types/src/memo.rs:289]: Check the error type

File: src/many-types/src/memo.rs [hunk 21]
```diff
@@ -0,0 +1,365 @@
...
+    fn mixed_decode_ok() {
+        let data = hex::encode(vec![1u8; MEMO_DATA_DEFAULT_MAX_SIZE]);
+        let cbor = format!(r#" [ "", h'{data}', "" ] "#);
+        let bytes = cbor_diag::parse_diag(cbor).unwrap().to_bytes();
+
+        assert!(minicbor::decode::<Memo>(&bytes).is_ok());
+    }
+
+    #[test]
+    fn mixed_decode_data_too_lare() {
```
Comment (fmorency) [inline src/many-types/src/memo.rs:284]:
```suggestion
    fn mixed_decode_bytes_too_lare() {
```

File: src/many-types/src/memo.rs [hunk 22]
```diff
@@ -0,0 +1,365 @@
...
+        let data = hex::encode(vec![1u8; MEMO_DATA_DEFAULT_MAX_SIZE + 1]);
+        let cbor = format!(r#" [ "", h'{data}', "" ] "#);
+        let bytes = cbor_diag::parse_diag(cbor).unwrap().to_bytes();
+
+        assert!(minicbor::decode::<Memo>(&bytes).is_err());
+    }
+
+    #[test]
+    fn mixed_decode_data_type_mismatch() {
```
Comment (fmorency) [inline src/many-types/src/memo.rs:293]:
```suggestion
    fn mixed_decode_memo_type_mismatch() {
```
nit

File: src/many-types/src/memo.rs [hunk 23]
```diff
@@ -0,0 +1,365 @@
...
+        let cbor = r#" 0 "#;
+        let bytes = cbor_diag::parse_diag(cbor).unwrap().to_bytes();
+
+        assert!(minicbor::decode::<Memo>(&bytes).is_err());
```
Comment (fmorency) [inline src/many-types/src/memo.rs:297]: Check the error type

File: src/many-types/src/memo.rs [hunk 24]
```diff
@@ -0,0 +1,365 @@
...
+    }
+
+    #[test]
+    fn mixed_decode_data_type_mismatch_array() {
```
Comment (fmorency) [inline src/many-types/src/memo.rs:301]:
```suggestion
    fn mixed_decode_memo_type_mismatch_array() {
```
nit

File: src/many-types/src/memo.rs [hunk 25]
```diff
@@ -0,0 +1,365 @@
...
+        let cbor = r#" [ "", 0, "" ] "#;
+        let bytes = cbor_diag::parse_diag(cbor).unwrap().to_bytes();
+
+        assert!(minicbor::decode::<Memo>(&bytes).is_err());
```
Comment (fmorency) [inline src/many-types/src/memo.rs:305]: Check the error type

**Review by fmorency — CHANGES_REQUESTED** [review]
Minor stuff. LGTM otherwise! Thanks!

File: src/many-modules/src/_4_events.rs (hunk 7 above)
Comment (hansl) [reply src/many-modules/src/_4_events.rs:964]: I don't want to be gloomy, but `default()` is how block 9400 came to be :/ I'm tempted to remove it from everywhere until we have a solid solution.

File: src/many-types/src/memo/legacy.rs (hunk 9 above)
Comment (hansl) [reply src/many-types/src/memo/legacy.rs]: The `Data` type was mixed with the old `Memo` type, so they're tied together. The new `Memo` type is essentially the merge of both of those two fields.

**Review by hansl — COMMENTED** [review]
PTAL.

**Review by fmorency — APPROVED** [review]
Lgtm

File: src/many-types/src/memo.rs [hunk 26]
```diff
@@ -131,6 +139,17 @@ impl<const M: usize> Memo<M> {
     }
 }
 
+// This helps comparisons.
+impl<const M: usize> PartialEq<str> for Memo<M> {
+    fn eq(&self, other: &str) -> bool {
+        if self.len() == 1 {
```
Comment (fmorency) [inline src/many-types/src/memo.rs:145]: Do we really need this check?

**Review by fmorency — APPROVED** [review]
LGTM assuming my question is addressed.

My macro-fu is unable to review this piece, but it looks like the others.

File: src/many-types/src/memo.rs (hunk 26 above)
Comment (hansl) [reply src/many-types/src/memo.rs:145]: Yes. If there's more than 1 string, `== "some string"` should always be false.

File: src/many-migration/src/lib.rs [hunk 27]
```diff
@@ -378,101 +139,300 @@ impl<'a, T, E> InnerMigration<'a, T, E> {
...
+        Self {
+            migration,
+            metadata,
+            enabled,
+            active: false,
+        }
+    }
+
+    /// Check the height and call the inner migration's methods.
+    pub fn apply_height(&mut self, storage: &mut T, block_height: u64) -> Result<(), E> {
```
Comment (fmorency) [inline src/many-migration/src/lib.rs:215]: nit: rename this method, it's confusing

File: src/many-migration/src/lib.rs [hunk 28]
```diff
@@ -378,101 +139,300 @@ impl<'a, T, E> InnerMigration<'a, T, E> {
...
+        if self.is_enabled() {
+            if block_height > self.metadata.block_height && !self.active {
```
Comment (fmorency) [inline src/many-migration/src/lib.rs:217]: `==` instead of `>`?

File: src/many-migration/src/lib.rs [hunk 29]
```diff
@@ -378,101 +139,300 @@ impl<'a, T, E> InnerMigration<'a, T, E> {
...
-                    "Migration {} is not of type `Regular`, skipping",
-                    self.name()
-                );
+                trace!("Skipping migration {}", self.name());
                 Ok(())
             }
         }
     }
 
     /// This function gets executed when the storage block height >= the migration block height
```
Comment (fmorency) [inline src/many-migration/src/lib.rs:153]: `>`

File: src/many-migration/src/lib.rs [hunk 30]
```diff
@@ -378,101 +139,300 @@ impl<'a, T, E> InnerMigration<'a, T, E> {
...
+        &self.inner[index.as_ref()]
+    }
+}
+
+pub fn load_migrations<'a, T, E>(
+    registry: &'a [InnerMigration<T, E>],
+    config: &str,
+) -> Result<MigrationSet<'a, T, E>, String> {
+    let config: MigrationConfig = serde_json::from_str(config).map_err(|e| e.to_string())?;
+    MigrationSet::load(registry, config, 0)
```
Comment (fmorency) [inline src/many-migration/src/lib.rs:411]: As discussed on Slack

File: src/many-migration/src/lib.rs [hunk 31]
```diff
@@ -378,101 +139,300 @@ impl<'a, T, E> InnerMigration<'a, T, E> {
...
+                MigrationType::Hotfix(_) => migration.disable(),
+                _ => migration.disable(),
+            }
+
+            (m.name.to_string(), migration)
         })
-        .collect()
+        .collect();
+
+    MigrationSet { inner, height: 0 }
```
Comment (fmorency) [inline src/many-migration/src/lib.rs:438]: As discussed on Slack

**Review by fmorency — CHANGES_REQUESTED** [review]
Minor stuff. Else LGTM

File: src/many-migration/src/lib.rs (hunk 28 above)
Comment (hansl) [reply src/many-migration/src/lib.rs:242]: How did that past the tests :/ Need to test more.

File: src/many-migration/src/lib.rs (hunk 29 above)
Comment (hansl) [reply src/many-migration/src/lib.rs:166]: Nope, when it's ==.

File: src/many-migration/src/lib.rs (hunk 30 above)
Comment (hansl) [reply src/many-migration/src/lib.rs:446]: That's the current height. This function is deprecated anyway.

File: src/many-migration/src/lib.rs (hunk 29 above)
Comment (fmorency) [reply src/many-migration/src/lib.rs:195]: This comment is still wrong

File: src/many-migration/src/lib.rs [hunk 32]
```diff
@@ -362,117 +164,396 @@ impl<T, E> InnerMigration<'static, T, E> {
...
+
+impl<T: IntoIterator<Item = impl Into<SingleMigrationConfig>>> From<T> for MigrationConfig {
+    fn from(value: T) -> Self {
+        Self(value.into_iter().map(Into::into).collect())
+    }
+}
+
+pub struct MigrationSet<'a, T: 'a, E: 'a = many_error::ManyError> {
+    inner: BTreeMap<String, Migration<'a, T, E>>,
+    height: u64,
```
Comment (fmorency) [inline src/many-migration/src/lib.rs:405]: Why are you keeping height here?

You don't use it anywhere but in `self.height = block_height;`

File: src/many-migration/src/lib.rs [hunk 33]
```diff
@@ -362,117 +164,396 @@ impl<T, E> InnerMigration<'static, T, E> {
...
+impl<T, E> From<(&InnerMigration<T, E>, Metadata)> for SingleMigrationConfig {
+    fn from((migration, metadata): (&InnerMigration<T, E>, Metadata)) -> Self {
+        Self {
+            name: migration.name.to_string(),
+            metadata,
+        }
+    }
+}
+
+#[derive(Default, Serialize, Deserialize)]
```
Comment (fmorency) [inline src/many-migration/src/lib.rs:375]: Please test serialization/deserialization of `MigrationConfig`

File: src/many-migration/src/lib.rs [hunk 34]
```diff
@@ -362,117 +164,396 @@ impl<T, E> InnerMigration<'static, T, E> {
...
+    fn initialize(&self, storage: &mut T) -> Result<(), E> {
         match &self.r#type {
             MigrationType::Regular(migration) => (migration.initialize_fn)(storage),
+            MigrationType::Hotfix(_) => Ok(()),
             _ => {
-                tracing::trace!(
-                    "Migration {} is not of type `Regular`, skipping",
-                    self.name()
-                );
+                trace!("Skipping migration {}", self.name());
```
Comment (fmorency) [inline src/many-migration/src/lib.rs:189]: Modify this message to be consistent with the others.

File: src/many-migration/src/lib.rs [hunk 35]
```diff
@@ -362,117 +164,396 @@ impl<T, E> InnerMigration<'static, T, E> {
...
-            let (&k, &v) = registry
-                .get_key_value(io.r#type)
-                .ok_or_else(|| format!("Unsupported migration type {}", io.r#type))?;
-            Ok((k, Migration::new(v, io.metadata, Status::Enabled)))
-        })
-        .collect::<Result<BTreeMap<_, _>, String>>()?
-        .into_iter()
-        .collect())
+impl MigrationConfig {
+    pub fn with_migration<T, E>(self, migration: &InnerMigration<T, E>) -> Self {
```
Comment (fmorency) [inline src/many-migration/src/lib.rs:380]: Test those

File: src/many-migration/src/lib.rs [hunk 36]
```diff
@@ -362,117 +164,396 @@ impl<T, E> InnerMigration<'static, T, E> {
...
+            .filter(|m| m.is_hotfix() && m.name() == name)
+        {
+            if let Some(r) = migration.hotfix(b, block_height) {
+                return Ok(Some(r));
+            }
+        }
+        Ok(None)
+    }
+
+    #[inline]
```
Comment (fmorency) [inline src/many-migration/src/lib.rs:483]: Test those

File: src/many-migration/src/lib.rs [hunk 37]
```diff
@@ -362,117 +164,396 @@ impl<T, E> InnerMigration<'static, T, E> {
...
+                Ok((config.name, Migration::new(v, config.metadata)))
+            })
+            .collect::<Result<BTreeMap<_, _>, String>>()?
+            .into_iter()
+            .collect();
+
+        // Activate all already active migrations. Do not call initialize though.
+        for v in inner.values_mut().filter(|m| m.is_enabled()) {
+            if height >= v.metadata.block_height {
+                v.active = true;
```
Comment (fmorency) [inline src/many-migration/src/lib.rs:453]: This line doesn't seem tested

**Review by fmorency — CHANGES_REQUESTED** [review]
Some questions/changes.

Do you have time to add more tests to memo, e.g., try_from(Either), Encode(bstr), try_from_iter(), push_str, push_bytevec, etc

File: src/many-migration/src/lib.rs (hunk 32 above)
Comment (hansl) [reply src/many-migration/src/lib.rs:405]: Removed.

File: src/many-migration/src/lib.rs (hunk 34 above)
Comment (hansl) [reply src/many-migration/src/lib.rs:189]: Actually the others are wrong, fixed.

File: src/many-migration/src/lib.rs (hunk 33 above)
Comment (hansl) [reply src/many-migration/src/lib.rs:375]: really?

Comment (fmorency) [reply src/many-migration/src/lib.rs:375]: Perhaps not. You seem to test it in fw anyway

Comment (fmorency) [reply src/many-migration/src/lib.rs:375]: Yes, I agree

**Comment by hansl** [comment]
I hope this is enough ;)

image

**Review by fmorency — APPROVED** [review]
Assuming CI is fixed

**Comment by fmorency** [comment]
Please log an issue with the remaining stuff to do/fix

**Pull request merged**